src/runtime-rs/crates/hypervisor/src/dragonball/inner.rs

```diff
@@ -374,13 +383,111 @@ impl DragonballInner {
         Ok((old_vcpus, new_vcpus))
     }
 
+    pub(crate) fn resize_memory(
+        &mut self,
+        old_mem_mb: u32,
+        new_mem_mb: u32,
+    ) -> Result<(u32, MemoryConfig)> {
+        // check the invalid request memory
+        if new_mem_mb > self.hypervisor_config().memory_info.default_maxmemory
+            && self.hypervisor_config().memory_info.default_maxmemory > 0
+        {
+            warn!(
+                sl!(),
+                "memory size unchanged, the request memory size is greater than the max memory {}",
+                self.hypervisor_config().memory_info.default_maxmemory
+            );
+
+            return Ok((
+                old_mem_mb,
+                MemoryConfig {
+                    ..Default::default()
+                },
+            ));
+        }
+
+        let had_mem_mb = self.config.memory_info.default_memory + self.mem_hotplug_size_mb;
+        match new_mem_mb.cmp(&had_mem_mb) {
+            Ordering::Greater => {
+                // clean virtio-ballon device before hotplug memory, resize to 0
```

Comment:
As old_mem_mb has passed to this function, I think ballon_size can be calculated out. I think we could write:
```rust
let balloon_size = had_mem_mb - old_mem_mb;
if balloon_size > 0 {
	insert ballon device
}
```

By this way, we can reduce one VMMAction.

Reply:
Nice catch! fixed